sw/simulator/nps/nps_fdm_gazebo.cpp

```diff
@@ -359,37 +360,50 @@ static void init_gazebo(void)
-  range_joint->GetAttribute("type")->Set("fixed");
-  range_joint->GetElement("parent")->Set("chassis");
-  range_joint->GetElement("child")->Set("range_sensors::base");
-#endif
-  // bebop front camera
-#ifdef NPS_SIMULATE_MT9F002
+#if NPS_SIMULATE_VIDEO
+  // Cameras
   sdf::ElementPtr link = vehicle_sdf->Root()->GetFirstElement()->GetElement("link");
   while (link) {
-    if (link->Get<string>("name") == "front_camera") {
+    if (link->Get<string>("name") == "front_camera" && link->GetElement("sensor")->Get<string>("name") == "mt9f002") {
+      float hfov = link->GetElement("sensor")->GetElement("camera")->GetElement("horizontal_fov")->Get<float>();
       int w = link->GetElement("sensor")->GetElement("camera")->GetElement("image")->GetElement("width")->Get<int>();
-      link->GetElement("sensor")->GetElement("camera")->GetElement("image")->GetElement("width")->Set(
-        w * MT9F002_OUTPUT_SCALER);
       int h = link->GetElement("sensor")->GetElement("camera")->GetElement("image")->GetElement("height")->Get<int>();
-      link->GetElement("sensor")->GetElement("camera")->GetElement("image")->GetElement("height")->Set(
-        h * MT9F002_OUTPUT_SCALER);
+
+      ignition::math::Pose3d pose = link->GetElement("pose")->Get<ignition::math::Pose3d>();
+      pose += ignition::math::Pose3d(0,0,0,0,-MT9F002_OFFSET_X * hfov,-MT9F002_OFFSET_Y * hfov * h / w);
+      link->GetElement("pose")->Set(pose);
+
+      link->GetElement("sensor")->GetElement("camera")->GetElement("horizontal_fov")->Set(hfov * MT9F002_OUTPUT_WIDTH / MT9F002_OUTPUT_HEIGHT / MT9F002_ZOOM);
+
       int env = link->GetElement("sensor")->GetElement("camera")->GetElement("lens")->GetElement("env_texture_size")->Get<int>();
-      link->GetElement("sensor")->GetElement("camera")->GetElement("lens")->GetElement("env_texture_size")->Set(
-        env * MT9F002_OUTPUT_SCALER);
-      cout << "Applied MT9F002_OUTPUT_SCALER (=" << MT9F002_OUTPUT_SCALER << ") to " << link->Get<string>("name") << endl;
-      link->GetElement("sensor")->GetElement("update_rate")->Set(MT9F002_TARGET_FPS);
-      cout << "Applied MT9F002_TARGET_FPS (=" << MT9F002_TARGET_FPS << ") to " << link->Get<string>("name") << endl;
+      link->GetElement("sensor")->GetElement("camera")->GetElement("lens")->GetElement("env_texture_size")->Set(env / MT9F002_ZOOM);
+
+      link->GetElement("sensor")->GetElement("camera")->GetElement("image")->GetElement("width")->Set(MT9F002_OUTPUT_WIDTH);
+      link->GetElement("sensor")->GetElement("camera")->GetElement("image")->GetElement("height")->Set(MT9F002_OUTPUT_HEIGHT);
+
+      if (MT9F002_TARGET_FPS){
+        int fps = Min(MT9F002_TARGET_FPS, link->GetElement("sensor")->GetElement("update_rate")->Get<int>());
+        link->GetElement("sensor")->GetElement("update_rate")->Set(fps);
+      }
+    } else if  (link->Get<string>("name") == "front_camera" && link->GetElement("sensor")->Get<string>("name") == "mt9v117") {
```

Comment:
"I think checking just the sensor name is already sufficient, same for the front camera."

Disagree, these defines should only be used if this particular sensor is flown, the ardrone or other vehicles will have different sensors with different settings.

Reply:
True, but the ardrone does not have an mt9f002 or mt9v117, right? I think testing these names is already sufficient, I do not see why checking "front_camera" helps as the same name is also used in other drones like the ardrone.

Reply:
Ah, I understood your original comment the other way around. I did it like this cause not all links have sensors and the check link->GetElement("sensor") would fail otherwise.

Reply:
Maybe GetSensorCount() is a better choice then, as it does not require specific link names for the cameras.

Reply:
I think that GetSensorCount is a gazebo thing and not an sdf thing. I don't think this is such a big issue given that the front_camera check was already there, I just augmented it to auto check the sensor type to get rid of the extra define.

Reply:
Oops, that is indeed a Gazebo thing. Let's keep it like this then, but change the `front_camera` to `bottom_camera` for the mt9v117.